app/controllers/protocols_controller.rb

```diff
@@ -586,17 +582,185 @@ def import
+    description_array.each do |e|
+      if e == 'created_on' && @json_object[e] && @json_object[e] != ''
+        new_e = e.slice(0, 1).capitalize + e.slice(1..-1)
+        new_e = new_e.tr('_', ' ')
+        description_string =
+          description_string + new_e.to_s + ':  ' +
+          Sanitize.clean(params['protocol']['created_at'].to_s) + "\n"
+      elsif e == 'tags' && @json_object[e].any? && @json_object[e] != ''
+        new_e = e.slice(0, 1).capitalize + e.slice(1..-1)
+        new_e = new_e.tr('_', ' ')
+        description_string = description_string + new_e.to_s + ': '
+        @json_object[e].each do |tag|
+          description_string =
+            description_string +
+            Sanitize.clean(tag['tag_name']) + ' , '
+        end
+        description_string += "\n"
+        # Since protocols description field doesnt show html,i just remove it
+        # because its even messier (using Sanitize)
+        # what this does is basically appends "FIELD NAME: "+" FIELD VALUE"
+        # to description for various fields
+      elsif @json_object[e] && @json_object[e] != ''
+        new_e = e.slice(0, 1).capitalize + e.slice(1..-1)
+        new_e = new_e.tr('_', ' ')
+        description_string +=
+          new_e.to_s + ':  ' +
+          Sanitize.clean(@json_object[e].to_s) + "\n"
+      end
+    end
+    @db_json['authors'] = params['protocol']['authors']
+    @db_json['created_at'] = params['protocol']['created_at']
+    @db_json['updated_at'] = params['protocol']['last_modified']
+    @db_json['description'] = description_string
+    @db_json['steps'] = {}
+    pos = -1
+    # these whitelists are there to not let some useless step components trough,
+    # that always have data set to null (data doesnt get imported over to json)
+    whitelist_simple = %w(1 6 17)
+    # (simple to map) id 1= step description, id 6= section (title),
+    # id 17= expected result
+    whitelist_complex = %w(8 9 15 18 19 20)
+    # (complex mapping with nested hashes) id 8 = software package,
+    # id 9 = dataset, id 15 = command, id 18 = attached sub protocol
+    # id 19= safety information ,
+    # id 20= regents (materials, like scinote samples kind of)
+    @json_object['steps'].each do |step| # loop over steps
+      pos += 1 # position of step (first, second.... etc),
+      @db_json['steps'][pos.to_s] = {} # the json we will insert into db
+      @db_json['steps'][pos.to_s]['position'] = pos
+      step['components'].each do |key, value|
+        # sometimes there are random index values as keys
+        # instead of hashes, this is a workaround to that buggy json format
+        key = value if value.class == Hash
+        if whitelist_simple.include?(key['component_type_id'])
+          # append is the string that we append values into for description
+          case key['component_type_id']
+          when '1'
+            @db_json = protocolsio_step_description_populate(@db_json, key, pos)
```

Comment:
All these three methods (protocolsio_step_description_populate, protocolsio_step_title_populate, protocolsio_step_expected_result_populate) are actually doing almost the same thing. It would be good, if you would combine them in just one method.